ocaml/xapi/helpers.ml

```diff
@@ -1234,3 +1234,67 @@ let host_supports_hvm ~__context host =
+      let statuses = List.filter_map (fun task -> try Some (Db.Task.get_status ~__context ~self:task) with _ -> None) tasks in
+      let unfinished = List.exists (fun state -> state = `pending) statuses in
+      if unfinished
+      then begin
+        let from = call_api_functions ~__context
+            (fun rpc session_id -> Client.Client.Event.from ~rpc ~session_id ~classes ~token ~timeout:30.0) in
+        debug "Using events to wait for tasks: %s" (String.concat "," classes);
+        let from = Event_types.event_from_of_rpc from in
+        process from.Event_types.token
+      end else
+        ()
+    in
+    process ""
+
+  let to_result ~__context ~of_rpc ~name ~t =
+    wait_for ~__context ~tasks:[t];
+    let fail msg = raise Api_errors.(Server_error (internal_error, [Ref.string_of t; Printf.sprintf "%s: %s"name msg])) in
+    let res =
+      match Db.Task.get_status ~__context ~self:t with
+      | `pending                 -> fail "task shouldn't be pending - we just waited on it"
+      | `cancelled | `cancelling -> raise Api_errors.(Server_error (task_cancelled, [Ref.string_of t]))
+      | `failure                 ->
+        begin match Db.Task.get_error_info ~__context ~self:t with
+        | [] | [_]     -> fail "couldn't extract error info from task"
+        | code::params -> raise (Api_errors.Server_error (code, params))
+        end
+      | `success                 ->
+        begin match Db.Task.get_result ~__context ~self:t with
+        | "" -> fail "nothing was written to task result"
+        | s  -> try Xmlrpc.of_string s |> of_rpc
+                with e -> fail (Printf.sprintf "result wasn't placed in task's (ref = %s) result field. error: %s" (Ref.string_of t) (Printexc.to_string e))
+        end
+      in
+    res
+end
+
+(** This function wraps internal async calls such that they can be used safely with versions of the API which do not support internal async (`old_api`'s).
+  * We achieve this by simply attempting to make the internal async call, and if the response is 'unknown method', then try the old, synchronous version.
+  * Callers of this function are expected to do the necessary error handling specific to their API call, so let all other exceptions propagate up. *)
+let try_internal_async ~__context
+                       ~(debug_tag : string) (* e.g. debug_tag="VM.migrate_send" *)
+                       (internal_async_fn : unit -> API.ref_task)
+                       (sync_fn : unit -> 'a) (* only used when we are talking to a version of the api without InternalAsync *)
+                       (marshaller : Rpc.t -> 'a) =
+  let res = try
+      let t = internal_async_fn () in `new_api t
+    with
+      Api_errors.Server_error(code, params) when code=Api_errors.message_method_unknown -> `old_api
+  in
+  match res with
+  | `old_api ->
+    D.info "%s: InternalAsync call was not known by destination - trying sync call instead" debug_tag;
```

Comment:
Do you really need to manually add the `debug_tag`? Since the calls are all wrapped in `Debug.with_thread_associated` (in server_helpers.ml), this should get included automatically. Also, you don't need the `D.` prefix, as the module is opened.